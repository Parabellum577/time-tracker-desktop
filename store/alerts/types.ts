import { ActionType } from 'typesafe-actions'

import * as actions from './actions'
import { IError } from '@services/types'

export type AlertsAction = ActionType<typeof actions>

export interface IAlert {
  alertType: 'success' | 'error' | 'warning' | 'coldstart'
  alertMessage: string | IError
}

export interface IAlertsState {
  alertsArray: IAlert[]
}
