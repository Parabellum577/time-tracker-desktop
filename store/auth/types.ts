import { ActionType } from 'typesafe-actions'
import { DeepReadonly } from 'utility-types'

import * as actions from './actions'
import { IError } from '@types'

export type AuthAction = ActionType<typeof actions>

export type IAuthState = DeepReadonly<{
  readonly authorized: boolean
  readonly error?: IError | null
  readonly loading: boolean
  readonly isLoadingData: boolean
}>
