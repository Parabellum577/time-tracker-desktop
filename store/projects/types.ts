import { ActionType } from 'typesafe-actions'

import { ISettingsItem, IProjectIssue, IProject } from '@types'
import * as actions from './actions'

export type IProjectsSettings = IProjectsSettingsItem[]

export interface IProjectsSettingsItem {
  project: string
  settings: ISettingsItem[]
}

export interface ISavedProjectIssues {
  project: string
  issues: IProjectIssue[]
}

export type IProjectsAction = ActionType<typeof actions>

export interface IProjectsState {
  readonly projectsSettings: IProjectsSettings
  readonly projectsIssues: ISavedProjectIssues[]
  readonly projects: IProject[]
}
