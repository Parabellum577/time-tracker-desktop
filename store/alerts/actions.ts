import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
