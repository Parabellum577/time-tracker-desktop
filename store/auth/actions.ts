import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
