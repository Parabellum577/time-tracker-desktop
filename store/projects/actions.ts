import { createAsyncAction, createAction } from 'typesafe-actions'
import { IError, IProjectIssue, IProject } from '@types'
import { IProjectsSettingsItem, ISavedProjectIssues } from './types'
import { IGetSettingsPayload } from '@services/api-types'

export const getProjectSettings = createAsyncAction(
  'user/GET_PROJECT_SETTINGS_REQUEST',
  'user/GET_PROJECT_SETTINGS_SUCCESS',
  'user/GET_PROJECT_SETTINGS_FAILURE',
)<IGetSettingsPayload, IProjectsSettingsItem, IError>()

export const handleProjectIssues = createAction('user/GET_PROJECT_ISSUES', action => (issues: ISavedProjectIssues) =>
  action(issues),
)

export const getProjects = createAsyncAction(
  'user/GET_PROJECTS_REQUEST',
  'user/GET_PROJECTS_SUCCESS',
  'user/GET_PROJECTS_FAILURE',
)<void, IProject[], IError>()

export const clearProjects = createAction('user/CLEAR_PROJECTS', action => () => action())
