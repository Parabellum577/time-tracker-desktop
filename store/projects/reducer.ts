import { getType } from 'typesafe-actions'
import * as _ from 'lodash'
import * as actions from './actions'
import { IProjectsAction, IProjectsState } from './types'

const initialState: IProjectsState = {
  projectsSettings: [],
  projectsIssues: [],
  projects: [],
}

export default (state: IProjectsState = initialState, action: IProjectsAction): IProjectsState => {
  switch (action.type) {
    case getType(actions.getProjectSettings.success):
      if (action.payload) {
        const filteredProjects = state.projectsSettings.filter(p => p.project !== action.payload.project)
        return {
          ...state,
          projectsSettings: [...filteredProjects, action.payload],
        }
      } else {
        return { ...state }
      }
    case getType(actions.handleProjectIssues):
      return {
        ...state,
        projectsIssues: [...state.projectsIssues.filter(p => p.project !== action.payload.project), action.payload],
      }
    case getType(actions.getProjects.success):
      const projects = action.payload

      return {
        ...state,
        projects,
      }
    case getType(actions.getProjects.failure):
      return { ...state }
    case getType(actions.clearProjects):
      return { ...state, projects: [] }
    default:
      return state
  }
}
