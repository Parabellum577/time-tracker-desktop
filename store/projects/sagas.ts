import { select, call, put, take, race } from 'redux-saga/effects'
import { getType, ActionType } from 'typesafe-actions'

import * as actions from './actions'
import * as loginActions from '@store/auth/actions'
import api from '@api'
import { ISettingsItem, IDefaultUserSettings, IProject, IProjectIssue, INamedID } from '@types'
import { IRootState } from '@store/rootReducer'
import { IProjectsSettingsItem } from './types'

const DEFAULT_SETTINGS: ISettingsItem[] = [
  { Key: 'track-apps', Value: 'apps' }, // ‘off’|’apps’|’all’
  { Key: 'screenshot-freq', Value: 'none' }, // 'none'|‘1x’|’2x’|’3x’|’smart’
  { Key: 'planning-max', Value: '24' }, // hours in String format
  { Key: 'manualReportLimit', Value: '4' }, // hours in String format
  { Key: 'track-idle-detection', Value: '10' }, // mins in String format
]

export const getUserSettings = (state: IRootState) => state.user.defaultUserSettings
export const getProjectSettingsFromStore = (state: IRootState) => state.projects.projectsSettings
export const getProjectFromStore = (state: IRootState) => state.projects.projects

export function* getProjectsSettingsSaga() {
  while (true) {
    const getProjectSettingsRequest: ActionType<typeof actions.getProjectSettings.request> = yield take(
      getType(actions.getProjectSettings.request),
    )

    try {
      let userSettings: IDefaultUserSettings
      let ProjectID = getProjectSettingsRequest.payload.ProjectID
      let TrackerID = getProjectSettingsRequest.payload.TrackerID
      if (TrackerID === 0 && ProjectID === '0') {
        userSettings = yield select(getUserSettings)
        TrackerID = userSettings.cold_settings.TrackerID
        ProjectID = userSettings.cold_settings.ProjectID
      }
      const projectKey = `${ProjectID}-${TrackerID}`
      const storedSettings: IProjectsSettingsItem[] = yield select(getProjectSettingsFromStore)
      let existSettings = storedSettings.find(i => i.project === projectKey)
      if (existSettings && !getProjectSettingsRequest.payload.hasUpdate) {
        if (userSettings) {
          existSettings = {
            ...existSettings,
            settings: [
              ...existSettings.settings.filter(s => s.Key !== 'track-idle-detection'),
              { Key: 'track-idle-detection', Value: userSettings.idle },
            ],
          }
        }
        yield put(actions.getProjectSettings.success(existSettings))
      } else {
        const settings: ISettingsItem[] = yield call(api.project.GetSettings, {
          ProjectID,
          TrackerID,
        })
        const allSettings = [...settings, ...DEFAULT_SETTINGS]
        // Remove duplicate settings, allow settings from project
        const uniqSettings = allSettings.filter((obj, pos, arr) => {
          return arr.map(mapObj => mapObj.Key).indexOf(obj.Key) === pos
        })
        yield put(actions.getProjectSettings.success({ project: projectKey, settings: uniqSettings }))
      }
    } catch (error) {
      console.error(error)
    }
  }
}

export function* getProjectsSaga() {
  while (true) {
    yield race({
      getProjects: take(getType(actions.getProjects.request)),
      loginUser: take(getType(loginActions.loginUser.success)),
    })
    try {
      const getProjects: IProject[] = yield call(api.project.GetProjects)
      yield put(actions.getProjects.success(getProjects))
      const storedSettings: IProject[] = yield select(getProjectFromStore)

      const coldProject = storedSettings.find(p => p.ColdProject)

      const issues: IProjectIssue[] = yield call(api.tracker.GetProjectIssues, {
        TrackerID: coldProject.TrackerID,
        ProjectID: coldProject.ProjectID,
      })
      const coldTask = issues.find(i => i.Title === 'Cold Task')
      yield put(actions.handleProjectIssues({ project: 'coldProject', issues: [coldTask] }))
    } catch (error) {
      console.error(error)
    }
  }
}
