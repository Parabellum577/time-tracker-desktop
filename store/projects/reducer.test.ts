import { ActionType, getType } from 'typesafe-actions'

import * as actions from './actions'
import { IProjectsState } from './types'
import projects from './reducer'

describe('[Store] Project Reducer', () => {
  it('[Success] [Snapshot] Should handle getProjectSettings', () => {
    const action: ActionType<typeof actions> = {
      type: getType(actions.getProjectSettings.success),
      payload: {
        project: 'project1',
        settings: [
          {
            Key: 'key1',
            Value: 'value2',
          },
        ],
      },
    }

    const currentState: IProjectsState = {
      projectsSettings: [],
      projectsIssues: [],
      projects: [],
    }

    expect(projects(currentState, action)).toMatchSnapshot()
  })

  it('[Failure] Should handle getProjectSettings', () => {
    const action: ActionType<typeof actions> = {
      type: getType(actions.getProjectSettings.failure),
      payload: {
        code: 666,
        message: 'SOME_ERROR',
      },
    }

    const currentState: IProjectsState = {
      projectsSettings: [],
      projectsIssues: [],
      projects: [],
    }

    expect(projects(currentState, action)).toMatchSnapshot()
  })
})
