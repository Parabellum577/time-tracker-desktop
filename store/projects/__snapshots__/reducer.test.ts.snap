// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] Project Reducer [Failure] Should handle getProjectSettings 1`] = `
Object {
  "projects": Array [],
  "projectsIssues": Array [],
  "projectsSettings": Array [],
}
`;

exports[`[Store] Project Reducer [Success] [Snapshot] Should handle getProjectSettings 1`] = `
Object {
  "projects": Array [],
  "projectsIssues": Array [],
  "projectsSettings": Array [
    Object {
      "project": "project1",
      "settings": Array [
        Object {
          "Key": "key1",
          "Value": "value2",
        },
      ],
    },
  ],
}
`;
